app/src/main/java/org/mozilla/fenix/home/WallpapersObserver.kt

```diff
@@ -28,18 +34,44 @@ import org.mozilla.fenix.wallpapers.WallpapersUseCases
  * when the [LifecycleOwner] is destroyed.
  *
  * @param appStore Holds the details abut the current wallpaper.
+ * @param settings Used for checking user's option for what wallpaper to use.
  * @param wallpapersUseCases Used for interacting with the wallpaper feature.
  * @param wallpaperImageView Serves as the target when applying wallpapers.
+ * @param backgroundWorkDispatcher Used for scheduling the wallpaper update when the state is updated
+ * with a new wallpaper.
  */
 class WallpapersObserver(
     private val appStore: AppStore,
+    private val settings: Settings,
     private val wallpapersUseCases: WallpapersUseCases,
     private val wallpaperImageView: ImageView,
+    backgroundWorkDispatcher: CoroutineDispatcher = Dispatchers.IO
 ) : DefaultLifecycleObserver {
     @VisibleForTesting
     internal var observeWallpapersStoreSubscription: Store.Subscription<AppState, AppAction>? = null
+
+    /**
+     * Coroutine scope for updating the wallpapers when an update is observed.
+     * Allows for easy cleanup when the client of this is destroyed.
+     */
+    @VisibleForTesting
+    internal var wallpapersScope = CoroutineScope(backgroundWorkDispatcher)
+
+    /**
+     * Setting the wallpaper assumes two steps:
+     * - load - running on IO
+     * - set - running on Main
+     * This property caches the result of [loadWallpaper] to be later used by [applyCurrentWallpaper].
+     */
+    @Volatile
     @VisibleForTesting
-    internal var wallpapersScope = CoroutineScope(Dispatchers.Main.immediate)
+    internal var currentWallpaperImage: Bitmap? = null
+
+    /**
+     * Listener for when the first observed wallpaper is loaded and available to be set.
+     */
+    @VisibleForTesting
+    internal val isWallpaperLoaded = CompletableDeferred<Unit>()
```

Comment:
Is there a clean way to combine these two properties, such that we could have `CompletableDeferred<Bitmap?>`?

Reply:
Can't say I'm happy with using all these properties and I did try to keep them to a minimum.

The `CompletableDeferred` is used just to be able to easily `await` having the first wallpaper. Otherwise we would maybe have to use a loop and a condition. This is more idiomatic and does not block the caller thread. 
It is publicly available but "completed" only once without passing other data to callers and allowing the `WallpaperObserver` to encapsulate all other state.

The `currentWallpaperImage` separate property helps decoupling the two operations - load bitmap and set bitmap through the use of this shared state. In my tests this allowed for quicker execution versus just one method with context switching between threads. By having two separate methods for load/set and being able to use a cached bitmap (same that's shown so there's no added memory pressure) we also reduce to a minimum the time needed to set the wallpaper after a configuration changes - we don't have to load again the bitmap.
It is a private property that is of no use to `WallpaperObserver` clients.